### PR TITLE
[Feature] Implemented /shield ability and related affects

### PR DESCRIPTION
**Pull request by KayenEQ**
8/10/21 
Better late than never! 

**Implemented the /shield ability.** 

/shield command mechanics
Warriors get this skill at level 30
Used by typing /shield while targeting a player
While active for the duration of 12 seconds baseline. The 'shield target' will take 50 pct less damage and the 'shielder' will be hit with the damage taken by the 'shield target' after all applicable mitigation is calculated.  The damage on the 'shielder' will be reduced by 25 percent, this reduction can be increased to 50 pct if equipping a shield. You receive a 1% increase in mitigation for every 2 AC on the shield. Shielder must stay with in a close distance (15 units) to your 'shield target'. If either move out of range, shield ends, no message given (how live does it). 	Both duration and shield range can be modified by AA. Recast is 3 minutes.

Example: Wizard is being hit by npc who normally is hitting for 100 max hit. Warrior shields Wizard. NPC now hits Wizard for 100 damage, since shielded, the Wizard takes 50 damage and the Warrior takes 75 damage (without a shield equipped)

**Perl quest function added that can be used to allow this ability to be used on any player/npc with maximum flexibility.**

`Perl_croak(aTHX_ "Usage: Mob::ShieldAbility(THIS, uint32 target_id, [int32 max_shielder_distance = 15], [int32 shield_duration = 12000], [int32 shield_target_mitigation= 50], [int32 shielder_mitigation = 50],  [bool use_aa = false], bool [can_shield_npc = true]");`

**Implemented SE_ExtendedShielding 230**
extends the range of your /shield ability by an amount of distance, base: distance units, limit: none, max: none

**Implemented SE_ShieldDuration 255**	
extends the duration of your /shield ability, base: seconds, limit: none, max: none

**Implemented SE_PetShield 210**
allows pet to 'shield' owner for 50 pct of damage taken for a duration 
base: Time multiplier 1=12 seconds, 2=24 ect), 
limit: mitigation on pet owner override (not on live) - Added to allow for more flexibility, if set to zero will get default behavior
max: mitigation on pet override (not on live) - Added to allow for more flexibility, if set to zero will get default behavior

Mages have this his AA in our database named 'Shared Health'. It will now be functional.

### CODE REVIEW

File: zone/perl_mob.cpp [hunk 1]
```diff
@@ -6301,6 +6263,50 @@ XS(XS_Mob_AddNimbusEffect) {
...
 }
 
+XS(XS_Mob_ShieldAbility); /* prototype to pass -Wmissing-prototypes */
+XS(XS_Mob_ShieldAbility) {
+	dXSARGS;
+	if (items < 2 || items > 6)
+		Perl_croak(aTHX_ "Usage: Mob::ShieldAbility(THIS, uint32 target_id, [int32 max_shielder_distance = 15], [int32 shield_duration = 12000], [int32 shield_target_mitigation= 50], [int32 shielder_mitigation = 50]"); // @categories Spells and Disciplines
+	{
+		Mob *THIS;
+		uint32             target_id = (uint32)SvUV(ST(1));
```
Comment (Akkadius) [inline zone/perl_mob.cpp:6273]: Alignment, formatting

File: zone/perl_mob.cpp [hunk 2]
```diff
@@ -6301,6 +6263,50 @@ XS(XS_Mob_AddNimbusEffect) {
...
+		int32			   max_shlder_distance;
```
Comment (Akkadius) [inline zone/perl_mob.cpp:6274]: can we just use the full word here; was confused as to what this was `shielder` I think is what you were going for

Comment (Akkadius) [reply zone/perl_mob.cpp:6274]: Same for vars below

File: zone/perl_mob.cpp [hunk 3]
```diff
@@ -6301,6 +6263,50 @@ XS(XS_Mob_AddNimbusEffect) {
...
+		int32			   shld_duration;
+		int32			   shld_target_mitigation;
+		int32			   shlder_mitigation;
+
+		VALIDATE_THIS_IS_MOB;
+		if (items < 3)
+			max_shlder_distance = 15;
+		else {
+			max_shlder_distance = max_shlder_distance = (int32)SvUV(ST(2));
```
Comment (Akkadius) [inline zone/perl_mob.cpp:6283]: These elses can just be declared as defaults above

Comment (Akkadius) [reply zone/perl_mob.cpp:6283]: Example

```cpp
XS(XS_Mob_ShieldAbility); /* prototype to pass -Wmissing-prototypes */
XS(XS_Mob_ShieldAbility) {
	dXSARGS;
	if (items < 2 || items > 6)
	Perl_croak(aTHX_ "Usage: Mob::ShieldAbility(THIS, uint32 target_id, [int32 max_shielder_distance = 15], [int32 shield_duration = 12000], [int32 shield_target_mitigation= 50], [int32 shielder_mitigation = 50]"); // @categories Spells and Disciplines
	{
		Mob    *THIS;
		uint32 target_id                = (uint32) SvUV(ST(1));
		int32  max_shielder_distance    = (int32) SvUV(ST(2));
		int32  shield_duration          = (int32) SvUV(ST(3));
		int32  shield_target_mitigation = (int32) SvUV(ST(4));
		int32  shielder_mitigation      = (int32) SvUV(ST(5));

		VALIDATE_THIS_IS_MOB;
		if (items < 3) {
			max_shlder_distance = 15;
		}

		if (items < 4) {
			shld_duration = 12000;
		}

		if (items < 5) {
			shld_target_mitigation = 50;
		}

		if (items < 6) {
			shlder_mitigation = 50;
		}

		THIS->ShieldAbility(target_id, max_shlder_distance, shld_duration, shld_target_mitigation, shlder_mitigation);

	}
	XSRETURN_EMPTY;
}
```

File: zone/perl_mob.cpp [hunk 4]
```diff
@@ -4179,44 +4179,6 @@ XS(XS_Mob_GetResist) {
 	XSRETURN(1);
 }
 
-XS(XS_Mob_GetShieldTarget); /* prototype to pass -Wmissing-prototypes */
```
Comment (Akkadius) [inline zone/perl_mob.cpp:4182]: There a reason we're removing calls here? If this is intended we should definitely make a PSA (Changelog) about this function being removed as some folks might be using it somewhere

File: zone/bonuses.cpp [hunk 5]
```diff
@@ -3541,6 +3552,34 @@ void Mob::ApplySpellsBonuses(uint16 spell_id, uint8 casterlevel, StatBonuses *ne
...
 				break;
 
+			case SE_ExtendedShielding:
+			{
+				if (AdditiveWornBonus) {
+					new_bonus->ExtendedShielding += effect_value;
+				}
+				else if (effect_value < 0 && new_bonus->ExtendedShielding > effect_value){
+					new_bonus->ExtendedShielding = effect_value;
+        }
```
Comment (Akkadius) [inline zone/bonuses.cpp:3562]: Crazy formatting here

File: zone/bonuses.cpp [hunk 6]
```diff
@@ -3541,6 +3552,34 @@ void Mob::ApplySpellsBonuses(uint16 spell_id, uint8 casterlevel, StatBonuses *ne
...
+			}
+
+			case SE_ShieldDuration:
+			{
+				if (AdditiveWornBonus) {
+					new_bonus->ShieldDuration += effect_value;
+				}
+				else if (effect_value < 0 && new_bonus->ShieldDuration > effect_value){
+					new_bonus->ShieldDuration = effect_value;
+        }
```
Comment (Akkadius) [inline zone/bonuses.cpp:3576]: Formatting here as well

File: zone/client_packet.cpp [hunk 7]
```diff
@@ -12806,87 +12806,101 @@ void Client::Handle_OP_SetTitle(const EQApplicationPacket *app)
...
+
+	//You are being shielded or already have a 'Shield Target'
+	if (GetShieldTargetID() || shield_target->GetShieldTargetID()) {
+		MessageString(Chat::White, ALREADY_SHIELDING);
+		return;
+	}
+	
+	//AA to increase SPA 230 extended shielding
+	int max_shlder_distance = 15;
+	max_shlder_distance += aabonuses.ExtendedShielding + itembonuses.ExtendedShielding + spellbonuses.ExtendedShielding;
```
Comment (Akkadius) [inline zone/client_packet.cpp:12880]: Can we full word the variables here as well for readability `max_shielder_distance`

**Comment by Akkadius** [comment]
Left a few comments; there's merge conflicts as well.

You're just pumping these out like the good ol days! <3

File: zone/perl_mob.cpp (hunk 4 above)
Comment (KayenEQ) [reply zone/perl_mob.cpp:4182]: The functions literally did nothing. The old shield variables had no functionality in game.

File: zone/bonuses.cpp (hunk 5 above)
Comment (KayenEQ) [reply zone/bonuses.cpp:3562]: merging in git web browser apparently doesn't work well.

File: zone/perl_mob.cpp (hunk 3 above)
Comment (KayenEQ) [reply zone/perl_mob.cpp]: Yours looks good, just replaced what I had with that.

File: zone/mob.h [hunk 8]
```diff
@@ -1133,13 +1131,28 @@ class Mob : public Entity {
...
+	inline uint32 GetShielderID() const { return shielder_id; }
+	inline void SetShielderID(uint32 val) { shielder_id = val; }
+	inline uint32 GetShieldTargetID() const { return shield_target_id; }
+	inline void SetShieldTargetID(uint32 val) { shield_target_id = val; }
+	inline int GetShieldTargetMitigation() const { return shield_target_mitigation; }
+	inline void SetShieldTargetMitigation(int val) { shield_target_mitigation = val; }
+	inline int GetShielderMitigation() const { return shielder_mitigation; }
+	inline void SetShielderMitigation(int val) { shielder_mitigation = val; }
+	inline int GetMaxShielderDistance() const { return shielder_max_distance; }
+	inline void SetShielerMaxDistance(int val) { shielder_max_distance = val; }
```
Comment (Akkadius) [inline zone/mob.h:1149]: Missing a `d` in `SetShielerMaxDistance`

File: zone/perl_mob.cpp (hunk 4 above)
Comment (Akkadius) [reply zone/perl_mob.cpp:4182]: Sounds good

**Comment by Akkadius** [comment]
We chatted async in DM's on some pointer feedback; will loop back after iteration

File: zone/spell_effects.cpp [hunk 9]
```diff
@@ -2950,6 +2950,20 @@ bool Mob::SpellEffect(Mob* caster, uint16 spell_id, float partial, int level_ove
 				break;
 			}
 
+			case SE_PetShield: {
+				if (IsPet()) {
+					Mob* petowner = GetOwner();
+					if (petowner) {
+
+						int shield_duration = spells[spell_id].base[i] * 12 * 1000;
+						int m_shield_target_mitigation = spells[spell_id].base2[i] ? spells[spell_id].base2[i] : 50;
```
Comment (Akkadius) [inline zone/spell_effects.cpp:2959]: This shouldn't prefix with `m_` as this is not a member variable of a class

File: zone/spell_effects.cpp [hunk 10]
```diff
@@ -2950,6 +2950,20 @@ bool Mob::SpellEffect(Mob* caster, uint16 spell_id, float partial, int level_ove
 				break;
 			}
 
+			case SE_PetShield: {
+				if (IsPet()) {
+					Mob* petowner = GetOwner();
+					if (petowner) {
+
```
Comment (Akkadius) [inline zone/spell_effects.cpp:2957]: Extra newline

File: zone/mob.cpp [hunk 11]
```diff
@@ -6196,6 +6197,131 @@ float Mob::GetDefaultRaceSize() const {
...
+			shielder->SetShielerMaxDistance(0);
+			shielder->shield_timer.Disable();
+		}
+		SetShielderID(0);
+		SetShieldTargetMitigation(0);
+	}
+
+	//If 'shielder' dies
+	if (GetShieldTargetID()) {
+
```
Comment (Akkadius) [inline zone/mob.cpp:6311]: Extra newline

File: zone/mob.cpp [hunk 12]
```diff
@@ -6196,6 +6197,131 @@ float Mob::GetDefaultRaceSize() const {
...
+	SetShielderMitigation(0);
+	SetShielerMaxDistance(0);
+	shield_timer.Disable();
+}
+
+void Mob::ShieldAbilityClearVariables()
+{
+	//If 'shield target' dies
+	if (GetShielderID()){
+
```
Comment (Akkadius) [inline zone/mob.cpp:6296]: Extra newline

File: zone/spell_effects.cpp (hunk 9 above)
Comment (Akkadius) [reply zone/spell_effects.cpp:2959]: image

**Comment by Akkadius** [comment]
Took care of a bunch of syntax and typo things

**Review by Akkadius — APPROVED** [review]

**Pull request merged**